core/src/toga/app.py

```diff
@@ -284,124 +304,91 @@ def __init__(
-            # folder.
+            # If the code is contained in appname/__main__.py, and you start the app
+            # using `python -m appname`, then __main__.__package__ will be "appname".
             try:
                 main_module_pkg = sys.modules["__main__"].__package__
-                if main_module_pkg == "":
-                    metadata_module_name = None
-                else:
-                    metadata_module_name = main_module_pkg
+                if main_module_pkg:
+                    distribution_name = main_module_pkg
             except KeyError:
-                # We use the existence of a __main__ module as a proxy for
-                # being in test conditions. This isn't *great*, but the __main__
-                # module isn't meaningful during tests, and removing it allows
-                # us to avoid having explicit "if under test conditions" checks.
-                # If there's no __main__ module, we're in a test, and we can't
-                # imply an app name from that module name.
-                metadata_module_name = None
-
-            # Try deconstructing the metadata module name from the app ID
-            if metadata_module_name is None and app_id:
-                metadata_module_name = app_id.split(".")[-1].replace("-", "_")
-
-            # If we still don't have a metadata module name, fall back to ``toga`` as a
-            # last resort.
-            if metadata_module_name is None:
-                metadata_module_name = "toga"
-
-        # Try to load the app metadata with our best guess of the module name.
+                # If there's no __main__ module, we're probably in a test.
+                pass
+
+        # Try deconstructing the distribution name from the app ID
+        if (distribution_name is None) and app_id:
+            distribution_name = app_id.split(".")[-1]
+
+        # If we still don't have a distribution name, fall back to ``toga`` as a
+        # last resort.
+        if distribution_name is None:
+            distribution_name = "toga"
+
+        # Try to load the app metadata with our best guess of the distribution name.
+        self._distribution_name = distribution_name
         try:
-            self.metadata = importlib_metadata.metadata(metadata_module_name)
-        except importlib_metadata.PackageNotFoundError:
+            self.metadata = importlib.metadata.metadata(distribution_name)
+        except importlib.metadata.PackageNotFoundError:
             self.metadata = Message()
 
-        # If the app name wasn't explicitly provided, look to the app metadata. If the
-        # metadata provides a "Name" key, use that as the app name; otherwise, fall back
-        # to the metadata module name (which might be "toga")
-        if app_name is None:
-            self._app_name = self.metadata.get("Name")
-            if self._app_name is None:
-                self._app_name = metadata_module_name
```

Comment:
I think it's better to stick to the simple rule that `distribution_name` is always the name by which the metadata was looked up. The previous behavior was never documented, and I can't think of a realistic use case for it.